Move all state to redux

probablycorey opened the pull request:
First, I've got four days to work on this for GitHub Hack Week, so I've abandoned some best practices in order to get some major progress done. This combines the ideas in https://github.com/cheshire137/gh-notifications-snoozer/pull/108 and https://github.com/cheshire137/gh-notifications-snoozer/pull/96 into one gigantic PR. To do this I had to get rid of some functionality, but now that there is less code it will be easier to add that functionality back in.

@cheshire137

Commit:
Add filter reducer

Commit:
Split reducer tests

Commit:
Add filter reducer

Commit:
Make filters more flexible

Commit:
Allow a filter to be selected

Commit:
Rename files

Commit:
Add reducers test

Commit:
Fix test

Commit:
Use array

Commit:
Remove only

Commit:
Remove Filter

Commit:
Use filter from redux

Commit:
Update EditFilter

Commit:
Make test pass

Commit:
Combine add and update

Commit:
Remove state from EditFilter

Commit:
Pass in filter

Commit:
Things be broke

Commit:
Fix filter list

Commit:
Fix EditFilter

Commit:
Make filter deletion work

Commit:
Get TaskList to work

Commit:
Add failing tests

Commit:
Make tests pass

Commit:
Moving things along

Commit:
Make this work with active filter

Commit:
Don’t show task list if there is no active filter

Commit:
Show task list even if there is no active filter

Commit:
Make filter switching work

Commit:
Use the active filter

Commit:
Single quotes

Commit:
Remove module

Commit:
Update tests and

Commit:
Remove models

Commit:
Remove module require

Commit:
Remove default filters

Commit:
Fix test

Commit:
:lipstick:

Commit:
THIS TEST SHALL PASS

Commit:
Add default filters back

Commit:
Store it all in userData again

Commit:
Use default filters

Commit:
Make the diff easier to read

cheshire137 committed:
Merge branch 'master' into move-filter-to-redux

cheshire137 committed:
Don't need 'else if' after return

cheshire137 committed:
Shorten long line

cheshire137 committed:
Need initial loadingTasks value

cheshire137 committed:
Use activeFilter prop for getTasksAfterNotifications

cheshire137 committed:
Wrap one-line return condition

Commit:
Send query

Commit:
Hack to make it work

Commit:
Remove pagination

Commit:
Remove task loading

Commit:
Install redux devtools

Commit:
Store tasks

Commit:
Remove electron config

Commit:
Update reducer styles

Commit:
Only show tasks based on filter

Commit:
Fix tests

Commit:
Fix test

Commit:
use correct function names

Commit:
Add simple-mock

cheshire137 commented:
> I had to get rid of some functionality

Can you list which functionality this branch removes?

cheshire137 commented:
/cc @summasmiff for 👀

cheshire137 commented on src/components/App/App.jsx:
```diff
@@ -34,37 +33,16 @@ class App extends React.Component {
-    if (user) {
-      const filters = new DefaultFilters(user.login)
-      filters.addDefaults()
+    if (this.props.filters.length === 0) {
```
Save two characters today, vote `< 1`! 🇺🇸

probablycorey replied:
While I'm a big fan of less code, I think the two chars make it more explicit in this specific case.

cheshire137 commented on src/components/EditFilter/EditFilter.jsx:
```diff
@@ -1,49 +1,28 @@
-      this.props.delete(this.props.filter)
-    }
-    this.props.save(key)
+    this.props.dispatch({ type: 'FILTERS_UPDATE', filter: this.props.filter })
```
Since you use filter a couple times, you could `const { filter } = this.props` at the beginning of the method.

probablycorey replied:
Since it is only used twice, I think this is ok for now. I did take your advice and do this in the render method, though.

cheshire137 commented on src/components/HiddenTaskList/HiddenTaskList.jsx:
```diff
@@ -70,14 +69,15 @@ class HiddenTaskList extends React.Component {
 
 HiddenTaskList.propTypes = {
   tasks: React.PropTypes.array.isRequired,
+  activeFilter: React.PropTypes.object.isRequired,
```
We hit problems before where some components expected an object for the filter and others expected just the string name. Are you consistent throughout this branch with expecting the filter to be an object?

probablycorey replied:
On the react side it is consistent. There is one spot on the redux side where it wasn't consistent, but I just committed a change to fix that.

cheshire137 commented on src/components/TaskList/TaskList.jsx:
```diff
@@ -70,37 +68,29 @@ class TaskList extends React.Component {
     }
   }
 
-  loadNextPage(event) {
```
We knew ye well, sweet pagination. 😿

probablycorey replied:
We will live in a pageless app soon! It will be glorious!

cheshire137 commented on src/components/TaskListItem/TaskListItem.jsx:
```diff
@@ -139,25 +135,9 @@ class TaskListItem extends React.Component {
-  repository: React.PropTypes.string.isRequired,
-  isPullRequest: React.PropTypes.bool.isRequired,
-  isSelected: React.PropTypes.bool,
-  isFocused: React.PropTypes.bool.isRequired,
```
Nice deletions, putting things in a task object!

cheshire137 commented on src/index.jsx:
```diff
@@ -1,16 +1,20 @@
 window.onload = function() {
-  const store = createStore(reducer, undefined, autoRehydrate())
-  const persistDir = remote.app.getPath('userData')
+  const reduxDevTools = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
```
Does this enable what you showed me in that gif?

probablycorey replied:
YES! It is amazing!

cheshire137 commented on src/main.js:
```diff
@@ -18,6 +19,12 @@ function createWindow() {
+    .then((name) => console.log(`Added Extension:  ${name}`))
+    .then(() => installExtension.default(installExtension.REDUX_DEVTOOLS)
+    .then((name) => console.log(`Added Extension:  ${name}`)))
+    .catch((err) => console.log('An error occurred: ', err))
```
You could drop the parentheses around these single arguments, like `(err)`.

probablycorey replied:
I'm going back and forth on this. I like the consistency of every function using params. But sometimes it is nice to leave them out. 

Let's pick a side and make our linter yell at us if we stray!

cheshire137 commented on src/models/Filter.js:
```diff
@@ -1,35 +0,0 @@
-'use strict'
```
Did Filter go away as a class because we're storing them in Redux now, so they need to be plain objects?

probablycorey replied:
Yeah. They are straight JS objects now.

cheshire137 commented on src/models/GitHub.js:
```diff
@@ -61,8 +59,8 @@ class GitHub extends Fetcher {
 
   // https://developer.github.com/v3/search/#search-issues
-  getTasks(query = Config.searchQuery) {
-    const params = `?per_page=${this.perPage}&q=${encodeURIComponent(query)}`
```
Why did pagination go away? Just to get tests passing and land this branch since it should be easy to add back later?

probablycorey replied:
I've got a scheme to make pagination unnecessary!

cheshire137 commented on src/models/GitHubAuth.js:
```diff
@@ -1,46 +1,25 @@
-    if (!isTest) {
-      return this.getToken() !== null
-    }
-    return storage.has(KEY)
```
I had a pickle of a time trying to get keytar to work in tests on Travis, I think it was. Did you get that working?

probablycorey replied:
I too had a pickle of a time. I was able to get it working by mocking it out in tests. I thought that approach made more sense because it keeps test code out of the app.

cheshire137 commented on src/reducers/FiltersReducer.js:
```diff
@@ -0,0 +1,58 @@
+const filtersUpdate = (existingFilters, { filter }) => {
+  let updatedFilters = null
+
+  if (!existingFilters.find(f => f.name === filter.name)) {
```
Can you swap the order of the if/else so we don't start with a negative condition?

cheshire137 commented on src/reducers/TasksReducer.js:
```diff
@@ -0,0 +1,143 @@
+  existingTasks.forEach(task => (tasksByKey[task.storageKey] = task))
+
+  // Update existingTasks with new values and add new existingTasks
+  tasks.forEach((newTask) => {
```
You can drop the parentheses.

cheshire137 reviewed (commented):
I'll try pulling this branch down locally and trying it out.

probablycorey commented:
> Can you list which functionality this branch removes?
1. I removed all calls to update notifications. Trying to keep those working and change how tasks are stored was difficult. But once we have the new redux interface kinks worked out I think it will be easy to add this functionality back in.
2. I removed the pagination because soon we'll be able to show all results without having `next page` in the app. I've already got this working in a follow-up PR!
3. I removed backward compatibility. Once we get to a spot where we are happy with the model structure we can write code to import a user's filters and tasks.
4. Filters don't automatically fetch tasks, you have to click the refresh button. Keeping this manual made it easier to debug. I plan on fetching new tasks when the app loads and resyncing every X minutes once everything is a little more stable.

cheshire137 commented:
> I removed backward compatibility. Once we get to a spot where we are happy with the model structure we can write code to import a user's filters and tasks.

Can you clarify what you mean by 'backward compatibility' here?

probablycorey commented:
> Can you clarify what you mean by 'backward compatibility' here?

When you start the app it will be like you started it for the first time ever. None of your filters, snoozes, archives or auth tokens will exist.

cheshire137 commented:
> When you start the app it will be like you started it for the first time ever. None of your filters, snoozes, archives or auth tokens will exist.

Oh that's fine!

Commit:
Pull out filter

Commit:
Call them filterQueries

Commit:
swap

Commit:
Update tests
